proposals/0021-vk-coop-matrix.md

```diff
@@ -0,0 +1,426 @@
+    following the mask cannot be represented. This is a limitation because we
+    cannot have variable arguments to `vk::ext_instruction` functions.
+
+### `vk::utils::ArithmeticSelector`
+
+The `vk::utils::ArithmeticSelector` class is a template class that takes a base
+type and has a series of static function that are implemented using inline
+SPIR-V. The functions can generate arithemtic operations for the base type or a
+vector of the base type.
+
+```c++
+template <class BaseType>
+class ArithmeticSelector {
+    template <class T> static T Negate(T a);
+    template <class T> static T Add(T a, T b);
+    template <class T> static T Sub(T a, T b);
+    template <class T> static T Mul(T a, T b);
+    template <class T> static T Div(T a, T b);
+};
+```
+
+There will be template specializations for the following types:
+
+*   `half`
+*   `float`
+*   `double`
+*   `int16_t`
+*   `uint16_t`
+*   `int32_t`
+*   `uint32_t`
+*   `int64_t`
+*   `uint64_t`
+
+### `vk::utils::ConversionSelector`
+
+The `vk::utils::ConversionSelector` class is a template class that can be used
+to generate SPIR-V instructions that will convert one numerical type to another.
+
+```c++
+// The conversion selector is will be used to convert one type to another
+// using the SPIR-V conversion instructions. See
+// https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_conversion_instructions.
+// SourceType and TargetType must be integer or floating point scalar type.
+template <class SourceType, class TargetType> class ConversionSelector {
+    // Converts an object of type S to an object of type T.
+    // S must be SourceType, a vector of SourceType, or a cooperative matrix
+    // of SourceType. T must be TargetType, a vector of TargetType, or a
+    // cooperative matrix of TargetType. T must have the same number of
+    // components as S. T is a cooperative matrix if and only if S is a
+    // cooperative matrix.
+    template <class T, class S> static T Convert(S a) {
+      return OpConvert<T>(a);
+    }
+  };
+```
+
+There will be template specializations for all pairs of the following types:
+
+*   `half`
+*   `float`
+*   `double`
+*   `int16_t`
+*   `uint16_t`
+*   `int32_t`
+*   `uint32_t`
+*   `int64_t`
+*   `uint64_t`
+
+### SPIR-V pointers
+
+To be able to pass a GroupShared array by reference, we introduce a new type and
+function to `vk/spirv.h`.
+
+~~~
+template <typename T>
+vk::WorkgroupSpirvPointer
+````
+
+This is a type with no members. An instance of this type can be created by calling
+
+~~~
+
+template <typename T> WorkgroupSpirvPointer<T>
+GetGroupSharedAddress([[vk::ext_reference]] T v); ```
+
+where `v` must be a object in GroupShared memory.
+
+For example,
+
+```
+groupshared float shared_data[64];
+...
+WorkgroupSpirvPointer<float> scalar_ptr = vk::GetGroupSharedAddress(shared_data[0]);
+WorkgroupSpirvPointer<float[64]> array_ptr = vk::GetGroupSharedAddress(shared_data);
```

Comment:
Are there any additional limitations on this WorkgroupSpirvPointer type? I'm concerned that use of this may lead to needing VariablePointer support (e.g. if these are loaded from a local variable).

Reply:
There might be some uses. I'll look into what is possible.

Reply:
I made the type `const`, which will stop the need for variable pointers in all reasonable situations. I have a test for on the DXC pr where noinline is used on a function with a pointer as a parameter. This would require variable pointers, and I show how to add it. I don't expect people to do that.

Otherwise, I expect the legalization pass to optimize away all function and private storage class variables.

If it is used on another storage class, the validator will complain with a poor error.